File: 670-maximum-swap/670-maximum-swap.js
/**
 * @param {number} num
 * @return {number}
 */
var maximumSwap = function (num) {
  const nums = num.toString().split("").map(x => +x);
  const n = nums.length;
  let [maxIdx, low, high] = [n - 1, 0, 0];

  for (let i = n - 1; i >= 0; i--) {
    const [curr, max] = [nums[i], nums[maxIdx]];
    if (curr > max) maxIdx = i;
    else if (curr < max) [low, high] = [i, maxIdx];
  }

  [nums[low], nums[high]] = [nums[high], nums[low]];

  return +nums.join("");
};

File: 670-maximum-swap/670-maximum-swap.py
class Solution:
    def maximumSwap(self, num: int) -> int:
        nums = list(str(num))
        n = len(nums)
        l, r, maxIdx = 0, 0, n - 1
        for i in range(n - 2, -1, -1):
            if int(nums[i]) > int(nums[maxIdx]):
                maxIdx = i
            elif int(nums[i]) < int(nums[maxIdx]):
                l, r = i, maxIdx
        nums[l], nums[r] = nums[r], nums[l]
        return int(''.join(nums))